lib/membrane_audio_interleaver.ex

```diff
@@ -0,0 +1,283 @@
+    if state.finished do
+      # end of stream already sent
+      {:ok, state}
+    else
+      sample_size = Caps.sample_size(caps)
+
+      state =
+        case Bunch.Access.get_in(state, [:pads, pad]) do
+          %{queue: queue} when byte_size(queue) < sample_size ->
+            %{state | finished: true}
+
+          _state ->
+            Bunch.Access.update_in(
+              state,
+              [:pads, pad],
+              &%{&1 | stream_ended: true}
+            )
+        end
+
+      if state.finished do
+        {{:ok, end_of_stream: :output}, state}
+      else
+        interleave_and_return(state)
+      end
+    end
+  end
+
+  @impl true
+  def handle_event(pad, event, _context, state) do
+    Membrane.Logger.debug("Received event #{inspect(event)} on pad #{inspect(pad)}")
+
+    {:ok, state}
+  end
+
+  @impl true
+  def handle_process(
+        pad,
+        %Buffer{payload: payload} = _buffer,
+        _context,
+        %{caps: caps} = state
+      ) do
+    if state.finished do
+      {:ok, state}
+    else
+      {new_queue_size, state} = add_payload(payload, pad, state)
+
+      if new_queue_size >= Caps.sample_size(caps) do
+        interleave_and_return(state)
+      else
+        {{:ok, redemand: :output}, state}
+      end
+    end
+  end
+
+  @impl true
+  def handle_caps(pad, caps, _context, state) do
+    case state.caps do
+      nil ->
+        state = %{state | caps: caps}
+        {{:ok, caps: {:output, %{caps | channels: 2}}, redemand: :output}, state}
+
+      ^caps ->
+        {:ok, state}
+
+      _invalid_caps ->
+        raise(
+          RuntimeError,
+          "received invalid caps on pad #{inspect(pad)}, expected: #{inspect(state.caps)}, got: #{inspect(caps)}"
+        )
+    end
+  end
+
+  # send demand to input pads where current queue is not long enough
+  defp do_handle_demand(size, %{pads: pads} = state) do
+    if state.finished do
+      {:ok, state}
+    else
+      demands =
+        Enum.map(
+          pads,
+          fn {pad, %{queue: queue}} ->
+            demand_size =
+              queue
+              |> byte_size()
+              |> then(&max(0, size - &1))
+
+            {:demand, {pad, demand_size}}
+          end
+        )
+
+      {{:ok, demands}, state}
+    end
+  end
+
+  # try to interleave channels and formulate proper element callback return message
+  defp interleave_and_return(state) do
+    case(try_interleave(state)) do
+      :none -> {:ok, state}
+      {:finished, {buffer, state}} -> {{:ok, buffer: buffer, end_of_stream: :output}, state}
+      {:ok, {buffer, state}} -> {{:ok, buffer: buffer}, state}
+    end
+  end
+
+  # interleave channels only if all queues are long enough (have at least 'sample_size' size)
+  defp try_interleave(%{caps: caps, pads: pads, order: order} = state) do
+    sample_size = Caps.sample_size(caps)
+
+    min_length =
+      min_queue_length(pads)
+      |> trunc_to_whole_samples(sample_size)
```

Comment:
It is a good practice to avoid using pipe operator just once

```suggestion
    min_length =
      pads
      |> min_queue_length()
      |> trunc_to_whole_samples(sample_size)
```

Reply:
fixed